src/plugins/interactive_setup/server/plugin.ts

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { Subscription } from 'rxjs';
+
+import type { TypeOf } from '@kbn/config-schema';
+import type { CorePreboot, Logger, PluginInitializerContext, PrebootPlugin } from 'src/core/server';
+
+import { ElasticsearchConnectionStatus } from '../common';
+import type { ConfigSchema, ConfigType } from './config';
+import { defineRoutes } from './routes';
+
+export class UserSetupPlugin implements PrebootPlugin {
+  readonly #logger: Logger;
+
+  #configSubscription?: Subscription;
+  #config?: ConfigType;
+  readonly #getConfig = () => {
+    if (!this.#config) {
+      throw new Error('Config is not available.');
+    }
+    return this.#config;
+  };
+
+  #elasticsearchConnectionStatus = ElasticsearchConnectionStatus.Unknown;
+  readonly #getElasticsearchConnectionStatus = () => {
+    return this.#elasticsearchConnectionStatus;
+  };
+
+  constructor(private readonly initializerContext: PluginInitializerContext) {
+    this.#logger = this.initializerContext.logger.get();
+  }
+
+  public setup(core: CorePreboot) {
+    this.#configSubscription = this.initializerContext.config
+      .create<TypeOf<typeof ConfigSchema>>()
+      .subscribe((config) => {
+        this.#config = config;
+      });
+
+    // We shouldn't activate interactive setup mode if we detect that user has already configured
+    // Elasticsearch connection manually: either if Kibana system user credentials are specified or
+    // user specified non-default host for the Elasticsearch.
+    // User can also set `interactiveSetup.forceSetup` config to `true` to force interactive setup mode.
+    const shouldActiveSetupMode =
+      this.#getConfig().forceSetup ||
+      (!core.elasticsearch.config.credentialsSpecified &&
+        core.elasticsearch.config.hosts.length === 1 &&
+        core.elasticsearch.config.hosts[0] === 'http://localhost:9200');
+    if (!shouldActiveSetupMode) {
+      this.#logger.debug(
+        'Interactive setup mode will not be activated since Elasticsearch connection is already configured.'
+      );
+      return;
+    }
+
+    let completeSetup: (result: { shouldReloadConfig: boolean }) => void;
+    core.preboot.holdSetupUntilResolved(
+      'Validating Elasticsearch connection configuration…',
+      new Promise((resolve) => {
+        completeSetup = resolve;
+      })
+    );
+
+    // If preliminary check above indicates that user didn't alter default Elasticsearch connection
```

Comment:
__note:__ we need to figure out what request to ES we need exactly, but I wanted to already have something simple so that we don't forget about that.